subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -245,10 +252,31 @@ public WorkValidationContext getValidationContext() {
     }
 
     private CachingResult executeWithNoEmptySources(UnitOfWork work, PreviousExecutionContext context) {
-        work.broadcastRelevantFileSystemInputs(false);
+        broadcastWorkInputs(work, false);
         return delegate.execute(work, context);
     }
 
+    private void broadcastWorkInputs(UnitOfWork work, boolean onlyPrimaryInputs) {
+        workInputListeners.broadcastFileSystemInputsOf(work, new CompositeFileCollection() {
+            @Override
+            public String getDisplayName() {
+                return work.getDisplayName() + " relevant file inputs";
+            }
+
+            @Override
+            protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
+                work.visitRegularInputs(new InputFingerprinter.InputVisitor() {
+                    @Override
+                    public void visitInputFileProperty(String propertyName, InputFingerprinter.InputPropertyType type, InputFingerprinter.FileValueSupplier value) {
+                        if (!onlyPrimaryInputs || type == InputFingerprinter.InputPropertyType.PRIMARY) {
+                            visitor.accept((FileCollectionInternal) value.getFiles());
+                        }
+                    }
+                });
+            }
+        });
+    }
```

Comment:
I think this is better. We could even go further, and hide `work` and the enum both, and just pass an opaque visitor here that the callee can use to visit whatever we deem important here. We don't have to do it in this PR though. Perhaps something to do when we restructure input visiting to use parameter objects.

Reply:
Yeah, we could hide more from the work here. We'd still need to expose the display name, though. Configuration caching also wanted to store the individual properties instead of a single file collection at some point, so that is also something to keep in mind when designing the opaque visitor.